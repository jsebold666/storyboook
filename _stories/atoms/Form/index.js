import React from 'react';
import { text, selectV2 as select, boolean, object } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';

import readme from './README.md';
import Form from '../../../components/atoms/Form';
import Validations from '../../../components/atoms/Validations';
import TextInput from '../../../components/atoms/TextInput';
import Submit from '../../../components/atoms/Submit';
import PasswordInput from '../../../components/atoms/PasswordInput/PasswordInput';

const style = { backgroundColor: '#f1f5f7', padding: '0 20px' };

let field1 = null;
let field2 = null;

const onSubmit = e => {
  console.log('Gone');
  action('submited');
};

const component = () => (
  <div style={style}>
    <br />
    <Form onSubmit={onSubmit}>
      <TextInput label="UsuÃ¡rio" validations={[Validations.required]} value={field1} />
      <br />
      <PasswordInput validations={[Validations.required]} label="Senha" value={field2} />
      <br />
      <TextInput label="Campo Opcional" value={field2} />
      <br />
      <Submit value="Submit" />
    </Form>
    <br />
  </div>
);

export default [readme, component];
