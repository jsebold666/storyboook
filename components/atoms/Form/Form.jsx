import React, { Component } from 'react';
import idGenerator from 'react-id-generator';

export const FormContext = React.createContext({
  register: () => {}
});
export const uid = idGenerator;

export default class Form extends Component {
  constructor() {
    super();
    this.validations = {};
    this.getChildValue = {};
    this.errors = {};
    this.updateChildState = {};
  }

  runValidations(id) {
    const validations = this.validations[id] || [];
    const value = this.getChildValue[id]();
    this.errors[id] = validations.map(validation => validation(value));
    this.updateChildState[id]({ validationError: this.errors[id][0] });
  }

  register(context) {
    const id = context.uid;
    this.validations[id] = context.props.validations || [];
    this.updateChildState[id] = context.setState.bind(context);
    this.getChildValue[id] = function() {
      return this.state.value;
    }.bind(context);
  }

  validateAll() {
    Object.keys(this.validations).forEach(id => {
      this.runValidations(id);
    });
  }

  countFieldsWithErrors() {
    return Object.keys(this.errors).filter(id => this.errors[id][0]).length;
  }

  onSubmit(e) {
    e.preventDefault();
    this.validateAll();
    if (this.countFieldsWithErrors() === 0) {
      console.log('Submiting form...');
      this.props.onSubmit && this.props.onSubmit(e);
    } else {
      console.log('Form has errors: ', this.errors);
    }
  }

  render() {
    return (
      <FormContext.Provider value={this}>
        <form {...this.props} onSubmit={this.onSubmit.bind(this)}>
          {this.props.children}
        </form>
      </FormContext.Provider>
    );
  }
}
