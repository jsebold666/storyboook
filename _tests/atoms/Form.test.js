/* globals mount */
import React from 'react';
import Form from '../../components/atoms/Form';

describe('Expect <Form>', () => {
  it('to render', () => {
    const props = {
      value: 'This is text',
      onChange: () => {}
    };

    const wrapper = mount(<Form {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('to render properly with empty', () => {
    const props = {
      value: '',
      onChange: () => {}
    };

    const wrapper = mount(<Form {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
