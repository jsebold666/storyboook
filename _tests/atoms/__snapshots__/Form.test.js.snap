// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <Form> to render 1`] = `
<Form
  onChange={[Function]}
  value="This is text"
>
  <form
    onChange={[Function]}
    onSubmit={[Function]}
    value="This is text"
  />
</Form>
`;

exports[`Expect <Form> to render properly with empty 1`] = `
<Form
  onChange={[Function]}
  value=""
>
  <form
    onChange={[Function]}
    onSubmit={[Function]}
    value=""
  />
</Form>
`;
